recipes-core/systemd/systemd_%.bbappend

```diff
@@ -20,3 +20,17 @@ SRC_URI += " ${@d.getVar('SYSTEMD_APPFW_PATCHES_' + d.getVar('PV', False)[0:3],
 
 # systemd-nspawn needs getent (from glibc).
 RDEPENDS_${PN} += "glibc-getent"
+
+# Separate systemd-networkd to a subpackage.
+PACKAGES_prepend     = "${PN}-networkd "
+FILES_${PN}-networkd = " \
+    ${base_libdir}/systemd/systemd-networkd \
+    ${base_libdir}/systemd/system/systemd-network.service \
+    ${base_libdir}/systemd/system/systemd-network.socket \
+"
+
+# And override service activation for the new subpackage.
+SYSTEMD_PACKAGES += "${PN}-networkd"
+SYSTEMD_SERVICE_${PN}-networkd = "systemd-networkd.socket"
+
+RDEPENDS_${PN} += "systemd-networkd"
```

Comment:
The split is not the main point here. systemd-networkd is still supposed to be pulled in always whenever systemd is installed, virtually undoing the split like you correctly pointed out.

The reason for the split is mild Yoctoism: this is the least intrusive way I could come up with to override SYSTEMD_SERVICE for systemd-networkd to be its socket file instead of its service file. IOW, you separate the component(s) you want to override the service file for into a subpackage, then specify the desired service file(s) as the SYSTEMD_SERVICE of the subpackage.